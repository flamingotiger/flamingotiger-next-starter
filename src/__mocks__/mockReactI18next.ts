/* global jest */
import React from 'react';
import * as reactI18next from 'react-i18next';

const hasChildren = (node: any) => node && (node.children || (node.props && node.props.children));

const getChildren = (node: any) => (node && node.children ? node.children : node.props && node.props.children);

const renderNodes = (reactNodes: any) => {
	if (typeof reactNodes === 'string') {
		return reactNodes;
	}

	return Object.keys(reactNodes).map((key, i) => {
		const child = reactNodes[key];
		const isElement = React.isValidElement(child);

		if (typeof child === 'string') {
			return child;
		}
		if (hasChildren(child)) {
			const inner: any = renderNodes(getChildren(child));
			return React.cloneElement(child, { ...child.props, key: i }, inner);
		}
		if (typeof child === 'object' && !isElement) {
			return Object.keys(child).reduce((str, childKey) => `${str}${child[childKey]}`, '');
		}

		return child;
	});
};

const useMock: any = [(k: string) => k, {}];
useMock.t = (k: string) => k;
useMock.i18n = {};

const { I18nextProvider } = reactI18next;
const { initReactI18next } = reactI18next;
const { setDefaults } = reactI18next;
const { getDefaults } = reactI18next;
const { setI18n } = reactI18next;
const { getI18n } = reactI18next;

module.exports = {
	// this mock makes sure any components using the translate HoC receive the t function as a prop
	withTranslation: jest.fn().mockImplementation(() => (Component: React.FC | React.ComponentClass) => {
		const ComponentCopy = { ...Component };
		ComponentCopy.defaultProps = { ...Component.defaultProps, t: () => '' };
		return ComponentCopy;
	}),
	Trans: ({ children }: { children: React.ReactNode }) => renderNodes(children),
	Translation: ({ children }: any) => children((k: string) => k, { i18n: {} }),
	useTranslation: () => useMock,

	// mock if needed
	I18nextProvider,
	initReactI18next,
	setDefaults,
	getDefaults,
	setI18n,
	getI18n
};
