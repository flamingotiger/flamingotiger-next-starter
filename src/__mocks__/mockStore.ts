import createMockStore, { MockStore } from 'redux-mock-store';
import * as ReactRedux from 'react-redux';
import store from '../store/index';

const mockStore = createMockStore([]);
const makeMockStore = (state = store.getState() || {}): MockStore<unknown> & { dispatch: {} } =>
	mockStore({ ...state });

export const mockUseReactRedux = () => {
	/* mocking useSelector on our mock store */
	const spyUseSelector = jest.spyOn(ReactRedux, 'useSelector');
	spyUseSelector.mockImplementation(() => store.getState());
	/* mocking useDispatch on our mock store  */
	const spyDispatch = jest.spyOn(ReactRedux, 'useDispatch');
	spyDispatch.mockImplementation(() => store.dispatch);
};

export default makeMockStore;
