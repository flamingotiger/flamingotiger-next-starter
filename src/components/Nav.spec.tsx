import React from 'react';
import { mount } from 'enzyme';
import Nav from './Nav';
import { mockUseReactRedux } from '../__mocks__/mockStore';

describe('<Nav/>', () => {
	beforeAll(() => {
		mockUseReactRedux();
	});
	it('match snapshot', () => {
		const wrapper = mount(<Nav />);
		expect(wrapper).toMatchSnapshot();
	});
});
