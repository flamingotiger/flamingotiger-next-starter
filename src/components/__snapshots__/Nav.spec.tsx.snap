// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> match snapshot 1`] = `
.c0 {
  color: #ffffff;
  text-align: center;
  line-height: 30px;
  font-size: 18px;
  font-weight: bold;
  margin: 0 10px;
  -webkit-transition: background-color 0.3s;
  transition: background-color 0.3s;
  border-radius: 5px;
}

.c0:active {
  background-color: #fbc02d;
}

.c0 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  background-color: #fdd835;
  height: 50px;
  min-width: 180px;
  text-align: center;
  line-height: 30px;
  font-size: 20px;
  font-weight: bold;
  color: white;
  margin: 0 10px;
  -webkit-transition: background-color 0.3s;
  transition: background-color 0.3s;
}

.c2:active {
  background-color: #fbc02d;
}

.c2 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 20px;
  margin-bottom: 20px;
}

<Nav>
  <nav
    style={
      Object {
        "margin": "auto",
        "width": "800px",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "flex-end",
          "margin": "10px 0",
        }
      }
    >
      <styled.div
        onClick={[Function]}
      >
        <div
          className="c0"
          onClick={[Function]}
        >
          <div>
            <img
              src="image"
            />
            <p />
          </div>
        </div>
      </styled.div>
    </div>
    <div
      style={
        Object {
          "border": "10px solid black",
        }
      }
    >
      <img
        alt="nextjs custom starter"
        src="/background_logo.jpg"
      />
    </div>
    <styled.ul>
      <ul
        className="c1"
      >
        <li>
          <styled.button>
            <button
              className="c2"
            >
              <Link
                href="/"
              >
                <a
                  href="/"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                >
                  List
                </a>
              </Link>
            </button>
          </styled.button>
          <styled.button
            onClick={[Function]}
          >
            <button
              className="c2"
              onClick={[Function]}
            >
              change-language
            </button>
          </styled.button>
        </li>
      </ul>
    </styled.ul>
  </nav>
</Nav>
`;
